text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,213 @@
+
+#### Collection endorsement policy
+
+Discovery can be used to get endorsing peers for a given chaincode where the transaction function will write to a given collection, and this will take into account the collection level endorsement policy.
+Ideally we could take care of these endorsement requirements in the Gateway without the client needing to have any implicit knowledge of them. Suggest extending contract metadata to include information on collections, which can be queried by Fabric Gateway. If that isn't possible then we might need some way for the client to specify the collections that will be used (not desirable).
+
+### State based endorsement
+
+This is driven by business logic and rules. Requires the client to supply constraints on endorsing organisations as part of the proposal endorsement.
+
+### Chaincode-to-chaincode endorsement
+
+Augment contract metadata for transactions with additional chaincodes / transactions that will be invoked, then we use discovery to narrow the set of endorsers.
+
+### Proposal response consistency checking
+
+Current client SDK implementations do minimal checking of proposal responses before sending to the orderer. There have been some requests for more rigorous checking of proposal responses prior to submit to the orderer.
+
+Prior to sending to the orderer, Fabric Gateway should check that proposal responses are:
+- Consistent, including having matching response payloads.
+- Meet all applicable endorsement requirements.
+
+### Ensure query results reflect latest ledger state
+
+Query results should be obtained from peers whose ledger height is at least that of the most recent events sent to clients.
+
+Scenario:
+- Client receipt of chaincode event triggers business process.
+- Business process queries ledger to obtain private data, which cannot be included in event payload since it is private.
+The private data must be consistent with the state of the ledger after commit of the transaction that emitted the chaincode event, so the query must be processed by a peer with ledger height of at least the block number that contained the chaincode event.
+
+Implementation options:
+- Only direct queries (and endorsement requests) to the peer(s) with the highest ledger height.
+- Client proposals include last observed ledger height so Gateway can target any peers with at least that ledger height.
+ A view of ledger height can be tracked by the Gateway from block events it has received.
+
+## Scaling and load balancing
+
+Currently a client application is responsible for load balancing its requests between multiple peers.  The Gateway will handle this on behalf of the set of currently connected client applications.
+
+Ideally, the Gateway itself will be stateless (i.e. not maintain client session state) allowing clients to be routed though an appriopriate load balancer.
```

Comment:
What should be the suggested strategy to provide high availability of the Gateway service?

Reply:
At the server end you could use a load balancer at the Fabric Gateway endpoint, which would transparently dispatch incoming requests across a set of Fabric Gateway instances.

At the client end you could use a load-balancing gRPC client connection when connecting the Gateway client, which would transparently dispatch requests across a set of Fabric Gateway endpoints.